### ISSUE TITLE
`deprecated` does not allow factory patterns.

### ISSUE THREAD

**Issue by phi-friday** [microsoft/pyright#8517]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

If you have a question about typing or a behavior that you’re seeing in Pyright (as opposed to a bug report or enhancement request), consider posting to the [Pyright discussion forum](https://github.com/microsoft/pyright/discussions).

**Describe the bug**
`pyright` allow factory patterns starting with `1.1.368`. https://github.com/microsoft/pyright/releases/tag/1.1.368
However, it does not support the following cases

I'm confused if this is a `pyright` issue or a `pylance` issue.

**Code or Screenshots**
```python
# pyright: strict
from __future__ import annotations

import warnings
from functools import wraps
from typing import TYPE_CHECKING, Any, TypeVar

if TYPE_CHECKING:
    import sys
    from collections.abc import Callable

    if sys.version_info >= (3, 13):
        deprecated = warnings.deprecated
    else:
        from typing_extensions import deprecated

_T = TypeVar("_T", bound="Callable[..., Any]")


def create_deprecated(text: str, *args: Any, **kwargs: Any) -> deprecated:  # noqa: D103
    text = f"{text} {args} {kwargs}"

    def custom_deprecated_func(func: _T) -> _T:
        @wraps(func)
        def inner(*inner_args: Any, **inner_kwargs: Any) -> Any:
            warnings.warn(text, DeprecationWarning, stacklevel=2)
            return func(*inner_args, **inner_kwargs)

        return inner  # pyright: ignore[reportReturnType]

    return custom_deprecated_func  # pyright: ignore[reportReturnType]


custom_deprecated: deprecated = create_deprecated("deprecated")


@custom_deprecated
def some_function() -> None:  # noqa: D103
    print("deprecated function")  # noqa: T201


some_function()

```
![스크린샷 2024-07-24 오후 6 15 54](https://github.com/user-attachments/assets/45516285-c964-47e6-85c4-b08b67342daf)

If your code relies on symbols that are imported from a third-party library, include the associated import statements and specify which versions of those libraries you have installed.

**VS Code extension or command-line**
Are you running pyright as a VS Code extension, a language server in another editor, integrated into Pylance, or the command-line tool? Which version?

vscode extensions:
```shell
❯ code --list-extensions --show-versions | grep python
donjayamanne.python-environment-manager@1.2.4
donjayamanne.python-extension-pack@1.7.0
kevinrose.vsc-python-indent@1.18.0
magicstack.magicpython@1.1.0
ms-python.debugpy@2024.8.0
ms-python.mypy-type-checker@2023.6.0
ms-python.python@2024.10.0
ms-python.vscode-pylance@2024.7.1
samwillis.python-inline-source@0.0.2
```

python version: `3.9.19`

**Comment by erictraut** [comment]
Pyright is working as intended here. This isn't a bug.

You can't use deprecated in this manner if you'd like it to work with a static type checker. Pyright doesn't actually run the code; it analyzes it statically. The `create_deprecated` function in your example completely obscures the deprecated message because it returns a plain `deprecated` object.

If you want `deprecated` to work in your editor, you'll need to avoid using such dynamic code. Your technique might work for runtime type checkers where the `create_deprecated` function is executed.

For more details about how type checkers and other static analysis tools are supposed to handle `deprecated`, refer to [this section](https://typing.readthedocs.io/en/latest/spec/directives.html#deprecated) of the Python typing spec. Pyright is conformant with the spec.
